sklearn/ensemble/_hist_gradient_boosting/grower.py

```diff
@@ -401,15 +400,38 @@ def _apply_shrinkage(self):
         for leaf in self.finalized_leaves:
             leaf.value *= self.shrinkage
 
-    def _initialize_root(self, gradients, hessians):
+    def _initialize_root(self):
         """Initialize root node and finalize it if needed."""
+        tic = time()
+        if self.interaction_cst is not None:
+            allowed_features = set().union(*self.interaction_cst)
+            allowed_features = np.fromiter(
+                allowed_features, dtype=np.uint32, count=len(allowed_features)
+            )
+            arbitrary_feature = allowed_features[0]
+        else:
+            allowed_features = None
+            arbitrary_feature = 0
+
+        # TreeNode init needs the total sum of gradients and hessians. Therefore, we
+        # first compute the histograms and then compute the total grad/hess on an
+        # arbitrary feature histogram. This way we replace a loop over n_samples by a
+        # loop over n_bins.
+        histograms = self.histogram_builder.compute_histograms_brute(
+            self.splitter.partition,  # =self.root.sample_indices
+            allowed_features,
+        )
+        self.total_compute_hist_time += time() - tic
+
+        tic = time()
         n_samples = self.X_binned.shape[0]
         depth = 0
-        sum_gradients = sum_parallel(gradients, self.n_threads)
+        histogram_array = np.asarray(histograms[arbitrary_feature])
+        sum_gradients = histogram_array["sum_gradients"].sum()
```

Comment:
What do you think about doing this?

```
        sum_gradients = sum_parallel_sum_gradients(
            histograms, arbitrary_feature, self.n_threads
        )
```

where we defined our own `sum_parallel_sum_gradients`:

```
def sum_parallel_sum_gradients(
        const hist_struct[:, ::1] hist,
        unsigned int arbitrary_feature,
        int n_threads,
):
    cdef:
        Y_DTYPE_C out = 0.
        unsigned int bin_idx

    for bin_idx in prange(
        hist.shape[1], schedule='static', nogil=True, num_threads=n_threads
    ):
        out += hist[arbitrary_feature, bin_idx].sum_gradients

    return out
```

Same for `sum_parallel_sum_hessians`.

---

It may be slightly faster because of OpenMP, but it's more code.

Reply:
The histogram of a feature has 256 bins. Therefore the openmp overhead is not worth it.